pkg/common/plugin/k8s/client/client.go

```diff
@@ -0,0 +1,68 @@
+package client
+
+import (
+	"github.com/kubernetes/client-go/kubernetes"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+// K8SClient is a client for querying k8s API server
+type K8SClient interface {
+	//GetNode returns the node name in which a given pod lives
+	GetNode(namespace, podName string) (string, error)
+}
+
+type k8sClient struct {
+	kubeConfigFilePath string
+	clientset          *kubernetes.Clientset
+}
+
+// NewK8SClient creates a new K8SClient.
+// There are two cases:
+// - If a kubeConfigFilePath is provided, config is taken from that file -> use for clients running out of a k8s cluster
+// - If not (empty kubeConfigFilePath), InClusterConfig is used          -> use for clients running in a k8s cluster
+func NewK8SClient(kubeConfigFilePath string) K8SClient {
+	return &k8sClient{
+		kubeConfigFilePath: kubeConfigFilePath,
+	}
+}
+
+func (c *k8sClient) GetNode(namespace, podName string) (string, error) {
+	// Reload config
+	err := c.reloadConfig()
+	if err != nil {
+		return "", err
+	}
+
+	// Get pod
+	pod, err := c.clientset.CoreV1().Pods(namespace).Get(podName, metav1.GetOptions{})
+	if err != nil {
+		return "", err
+	}
+
+	return pod.Spec.NodeName, nil
+}
+
+func (c *k8sClient) reloadConfig() error {
```

Comment:
either `reloadConfig` returns the clientset, or you need to provide a lock to protect races on `c.clientset`

Reply:
The lock isn't sufficiently protecting access (GetNode doesn't hold the lock when accessing `c.clientset`). Since the client is rebuilt on every request, there doesn't seem much value in holding onto that state in the struct. I think it's a bit cleaner to just return the client from `reloadConfig` (which could be renamed to `loadClient`)